File: backend/app/routers/locations.py
"""restaurant_location endpoints + hours/photos/managers sub-resources. See
docs/API_CONTRACTS.md "Locations (restaurant_location)" and "Location
Managers".

Public: GET /locations/{id} (backend/CLAUDE.md "Public Routes"). Every
other route requires auth, re-validated server-side on every write (never
the JWT claims alone) via the dependencies in `app/dependencies/auth.py`.
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies.auth import (
    CurrentUser,
    get_current_user_optional,
    require_location_owner_only,
    require_location_owner_or_admin,
    require_location_read_access,
    require_location_write_access,
    require_owner,
)
from app.dependencies.db import get_db
from app.schemas.hours import HoursReplaceRequest, HoursResponse
from app.schemas.location import (
    LocationCreate,
    LocationOut,
    LocationStatusUpdate,
    LocationUpdate,
)
from app.schemas.location_manager import (
    AssignManagerRequest,
    LocationManagerListResponse,
    LocationManagerOut,
)
from app.schemas.location_reopen import ReopenRequestCreate, ReopenRequestOut
from app.schemas.photo import PhotoCreate, PhotoOut, PhotoUpdate, UploadUrlRequest, UploadUrlResponse
from app.services import location_manager_service, location_reopen_service, location_service

router = APIRouter(prefix="/locations", tags=["locations"])


@router.get("/{location_id}", response_model=LocationOut)
async def get_location(
    location_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser | None = Depends(get_current_user_optional),
) -> LocationOut:
    # `current_user` (optional -- public by default) is what lets
    # `location_service.get_location` distinguish "hidden to everyone" from
    # "hidden to the public but visible to its own owner/admin/assigned
    # manager" — see that function's docstring and
    # `_caller_may_view_hidden_location`. Previously this route never passed
    # a caller through at all, so a hidden location 404'd even for its own
    # owner viewing the detail page while signed in; that's the bug this
    # closes (docs/PROJECT_PLAN.csv row for this task).
    return await location_service.get_location(db, location_id, current_user)


@router.post("", response_model=LocationOut, status_code=status.HTTP_201_CREATED)
async def create_location(
    body: LocationCreate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_owner),
) -> LocationOut:
    return await location_service.create_location(db, body, current_user)


@router.patch("/{location_id}", response_model=LocationOut)
async def update_location(
    location_id: int,
    body: LocationUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> LocationOut:
    return await location_service.update_location(db, location_id, body, current_user)


@router.delete("/{location_id}", status_code=status.HTTP_204_NO_CONTENT, response_model=None)
async def delete_location(
    location_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_owner_or_admin),
) -> None:
    await location_service.delete_location(db, location_id, current_user)


@router.post("/{location_id}/status", response_model=LocationOut)
async def update_location_status(
    location_id: int,
    body: LocationStatusUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_owner_or_admin),
) -> LocationOut:
    """Owner (or admin) self-service status change — see
    `location_service.update_location_status` for the one asymmetric rule
    (no self-service exit from `closed_pending_reopen`; use
    `POST /locations/{id}/reopen-requests` instead). No manager path,
    same as `DELETE /locations/{id}` — see `require_location_owner_or_admin`.
    """
    return await location_service.update_location_status(db, location_id, body.status, current_user)


@router.post(
    "/{location_id}/reopen-requests",
    response_model=ReopenRequestOut,
    status_code=status.HTTP_201_CREATED,
)
async def create_reopen_request(
    location_id: int,
    body: ReopenRequestCreate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_owner_only),
) -> ReopenRequestOut:
    """Owner submits a request to reopen a `closed_pending_reopen` location
    — the only path back to `active` for that status (see
    `app/services/location_reopen_service.create_reopen_request`).
    Owner-only, not admin (see `require_location_owner_only`) — an admin
    doesn't need to "request" anything, they can already
    `POST /location-reopen-requests/{id}/approve` directly, or in principle
    the future could get an admin-initiated reopen; out of scope here since
    the task only asked for owner submission + admin approve/reject.
    """
    location = await location_service.get_location_or_404(db, location_id)
    request = await location_reopen_service.create_reopen_request(db, location, current_user, body.notes)
    return location_reopen_service.to_reopen_request_out(request)


@router.put("/{location_id}/hours", response_model=HoursResponse)
async def replace_location_hours(
    location_id: int,
    body: HoursReplaceRequest,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> HoursResponse:
    return await location_service.replace_location_hours(db, location_id, body.hours, current_user)


@router.post("/{location_id}/photos/upload-url", response_model=UploadUrlResponse)
async def create_photo_upload_url(
    location_id: int,
    body: UploadUrlRequest,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> UploadUrlResponse:
    return await location_service.create_photo_upload_url(db, location_id, body.content_type)


@router.post(
    "/{location_id}/photos", response_model=PhotoOut, status_code=status.HTTP_201_CREATED
)
async def create_location_photo(
    location_id: int,
    body: PhotoCreate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> PhotoOut:
    return await location_service.create_location_photo(db, location_id, body, current_user)


@router.patch("/{location_id}/photos/{photo_id}", response_model=PhotoOut)
async def update_location_photo(
    location_id: int,
    photo_id: int,
    body: PhotoUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> PhotoOut:
    return await location_service.update_location_photo(db, location_id, photo_id, body)


@router.delete("/{location_id}/photos/{photo_id}", status_code=status.HTTP_204_NO_CONTENT, response_model=None)
async def delete_location_photo(
    location_id: int,
    photo_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_write_access),
) -> None:
    await location_service.delete_location_photo(db, location_id, photo_id)


@router.post(
    "/{location_id}/managers",
    response_model=LocationManagerOut,
    status_code=status.HTTP_201_CREATED,
)
async def assign_location_manager(
    location_id: int,
    body: AssignManagerRequest,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_owner_only),
) -> LocationManagerOut:
    location = await location_service.get_location_or_404(db, location_id)
    return await location_manager_service.assign_manager(
        db, location, current_user, body.manager_email
    )


@router.get("/{location_id}/managers", response_model=LocationManagerListResponse)
async def list_location_managers(
    location_id: int,
    active_only: bool = Query(default=False),
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_read_access),
) -> LocationManagerListResponse:
    # Forced true for a manager caller (docs/API_CONTRACTS.md "GET
    # /locations/{id}/managers" — a manager sees who currently manages
    # this location, not the full removal history); owner/admin get the
    # `active_only` query param as given, default false (full history).
    effective_active_only = active_only or current_user.role == "manager"
    await location_service.get_location_or_404(db, location_id)
    results = await location_manager_service.list_managers(
        db, location_id, effective_active_only
    )
    return LocationManagerListResponse(results=results)


@router.delete(
    "/{location_id}/managers/{manager_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_model=None,
)
async def remove_location_manager(
    location_id: int,
    manager_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_location_owner_or_admin),
) -> None:
    await location_manager_service.deactivate_manager(
        db, location_id, manager_id, current_user
    )

File: frontend/src/lib/api/locations.ts
// Typed client for /locations (restaurant_location) and its sub-resources
// (hours, photos, managers) — docs/API_CONTRACTS.md "Locations
// (`restaurant_location`)" and "Location Managers (`location_manager`)".
import { apiFetch } from "./client";
import type {
  AssignLocationManagerInput,
  CreateLocationInput,
  CreatePhotoInput,
  LocationCuisineTagsResponse,
  LocationDetail,
  LocationManager,
  LocationManagersResponse,
  Photo,
  PhotoUploadUrlInput,
  PhotoUploadUrlResponse,
  UpdateLocationCuisineTagsInput,
  UpdateLocationHoursInput,
  UpdateLocationHoursResponse,
  UpdateLocationInput,
  UpdateLocationStatusInput,
  UpdatePhotoInput,
} from "@/types/location";

/**
 * GET /locations/{id} — public by default, but caller-aware: a hidden
 * location (any status other than `active`) 404s for anyone without real
 * access, UNLESS `accessToken` is passed AND that caller is the owner,
 * admin, or an assigned manager (see `location_service.get_location` /
 * `_caller_may_view_hidden_location`, backend/app/services/location_service.py).
 * Pass `accessToken` whenever the caller is signed in and might legitimately
 * need to see their own hidden location (the portal location editor, the
 * owner/manager console's own tiles) — omit it for a genuinely public/SSR
 * read (the public restaurant page), where an anonymous 404 on a hidden
 * location is the correct, intended behavior.
 */
export async function getLocationById(
  id: number,
  accessToken?: string
): Promise<LocationDetail> {
  return apiFetch<LocationDetail>(
    `/locations/${id}`,
    { method: "GET" },
    { accessToken, revalidateSeconds: accessToken ? undefined : 60 }
  );
}

/**
 * POST /locations/{id}/status — auth: owner (owns parent brand) or admin,
 * no manager path (docs/PROJECT_PLAN.csv "Location status lifecycle";
 * backend/app/routers/locations.py). The one asymmetric rule (no
 * self-service exit from `closed_pending_reopen`) is enforced server-side —
 * this call 409s the same way any other invalid transition would.
 */
export async function updateLocationStatus(
  id: number,
  input: UpdateLocationStatusInput,
  accessToken: string
): Promise<LocationDetail> {
  return apiFetch<LocationDetail>(
    `/locations/${id}/status`,
    { method: "POST", body: JSON.stringify(input) },
    { accessToken }
  );
}

/** POST /locations — auth: owner (must own the parent brand). */
export async function createLocation(
  input: CreateLocationInput,
  accessToken: string
): Promise<LocationDetail> {
  return apiFetch<LocationDetail>(
    "/locations",
    { method: "POST", body: JSON.stringify(input) },
    { accessToken }
  );
}

/**
 * PATCH /locations/{id} — auth: owner (owns parent brand) or manager with
 * an active assignment for this location (checked server-side, never from
 * the JWT alone).
 */
export async function updateLocation(
  id: number,
  input: UpdateLocationInput,
  accessToken: string
): Promise<LocationDetail> {
  return apiFetch<LocationDetail>(
    `/locations/${id}`,
    { method: "PATCH", body: JSON.stringify(input) },
    { accessToken }
  );
}

/**
 * PUT /locations/{id}/cuisine-tags — auth: owner (owns parent brand), assigned
 * manager, or admin. Full replace of THIS location's cuisine/dietary tags
 * (tags are per location; docs/API_CONTRACTS.md "PUT /locations/{id}/cuisine-tags").
 */
export async function replaceLocationCuisineTags(
  id: number,
  input: UpdateLocationCuisineTagsInput,
  accessToken: string
): Promise<LocationCuisineTagsResponse> {
  return apiFetch<LocationCuisineTagsResponse>(
    `/locations/${id}/cuisine-tags`,
    { method: "PUT", body: JSON.stringify(input) },
    { accessToken }
  );
}

/**
 * DELETE /locations/{id} — auth: owner (owns parent brand) or admin.
 * Soft delete (sets is_active=false) — see docs/API_CONTRACTS.md.
 */
export async function deleteLocation(
  id: number,
  accessToken: string
): Promise<void> {
  return apiFetch<void>(
    `/locations/${id}`,
    { method: "DELETE" },
    { accessToken }
  );
}

/**
 * DELETE /locations/{id}/permanent — auth: owner (owns parent brand) or
 * admin. Real, irreversible row delete — NOT the soft-hide `deleteLocation`
 * above. Backend guardrails (docs/API_CONTRACTS.md "DELETE
 * /locations/{id}/permanent"): the location must already be non-`active`,
 * have no active manager assignment, and no pending claim or reopen
 * request — each surfaces as its own `409` with a distinct `code`, message
 * passed through as-is by the caller (see `removeLocationAction`).
 */
export async function removeLocationPermanently(
  id: number,
  accessToken: string
): Promise<void> {
  return apiFetch<void>(
    `/locations/${id}/permanent`,
    { method: "DELETE" },
    { accessToken }
  );
}

/** PUT /locations/{id}/hours — full week replacement. */
export async function updateLocationHours(
  id: number,
  input: UpdateLocationHoursInput,
  accessToken: string
): Promise<UpdateLocationHoursResponse> {
  return apiFetch<UpdateLocationHoursResponse>(
    `/locations/${id}/hours`,
    { method: "PUT", body: JSON.stringify(input) },
    { accessToken }
  );
}

/**
 * POST /locations/{id}/photos/upload-url — step 1 of the S3 presigned
 * upload flow (root CLAUDE.md media pattern: presigned URL, never through
 * Lambda). This is a presigned **POST** (not PUT) — the client submits
 * `upload_url` as a multipart form, every entry in `fields` as its own
 * form field plus the file itself under the field name `file` (S3's
 * presigned-POST convention; see docs/API_CONTRACTS.md "POST
 * /locations/{id}/photos/upload-url" for why: only presigned POST can
 * enforce the 5MB `content-length-range` cap). Then calls
 * `createLocationPhoto` below with the same `s3_key`.
 */
export async function getLocationPhotoUploadUrl(
  id: number,
  input: PhotoUploadUrlInput,
  accessToken: string
): Promise<PhotoUploadUrlResponse> {
  return apiFetch<PhotoUploadUrlResponse>(
    `/locations/${id}/photos/upload-url`,
    { method: "POST", body: JSON.stringify(input) },
    { accessToken }
  );
}

/** POST /locations/{id}/photos — step 2, records the row after the S3 upload succeeds. */
export async function createLocationPhoto(
  id: number,
  input: CreatePhotoInput,
  accessToken: string
): Promise<Photo> {
  return apiFetch<Photo>(
    `/locations/${id}/photos`,
    { method: "POST", body: JSON.stringify(input) },
    { accessToken }
  );
}

/** PATCH /locations/{id}/photos/{photo_id} — reorder or set as cover. */
export async function updateLocationPhoto(
  id: number,
  photoId: number,
  input: UpdatePhotoInput,
  accessToken: string
): Promise<Photo> {
  return apiFetch<Photo>(
    `/locations/${id}/photos/${photoId}`,
    { method: "PATCH", body: JSON.stringify(input) },
    { accessToken }
  );
}

/** DELETE /locations/{id}/photos/{photo_id} — hard row delete. */
export async function deleteLocationPhoto(
  id: number,
  photoId: number,
  accessToken: string
): Promise<void> {
  return apiFetch<void>(
    `/locations/${id}/photos/${photoId}`,
    { method: "DELETE" },
    { accessToken }
  );
}

/**
 * POST /locations/{id}/managers — auth: owner only (must own the parent
 * brand). No admin path, no manager-assigns-manager path — see
 * docs/API_CONTRACTS.md "POST /locations/{id}/managers".
 */
export async function assignLocationManager(
  id: number,
  input: AssignLocationManagerInput,
  accessToken: string
): Promise<LocationManager> {
  return apiFetch<LocationManager>(
    `/locations/${id}/managers`,
    { method: "POST", body: JSON.stringify(input) },
    { accessToken }
  );
}

/**
 * GET /locations/{id}/managers — auth: owner (owns parent brand), admin, or
 * a manager with an active assignment on this location. `activeOnly` is
 * forced true server-side for a manager caller regardless of what's passed
 * here (docs/API_CONTRACTS.md).
 */
export async function getLocationManagers(
  id: number,
  params: { activeOnly?: boolean } = {},
  accessToken: string
): Promise<LocationManagersResponse> {
  const query = params.activeOnly ? "?active_only=true" : "";
  return apiFetch<LocationManagersResponse>(
    `/locations/${id}/managers${query}`,
    { method: "GET" },
    { accessToken }
  );
}

/**
 * DELETE /locations/{id}/managers/{manager_id} — auth: owner (owns parent
 * brand) or admin. Soft-deactivate (is_active=false), idempotent — calling
 * again on an already-inactive row still returns 204.
 * `manager_id` is the assignment row's own PK (`location_manager.id`), not
 * the manager's `user_id`.
 */
export async function removeLocationManager(
  id: number,
  managerId: number,
  accessToken: string
): Promise<void> {
  return apiFetch<void>(
    `/locations/${id}/managers/${managerId}`,
    { method: "DELETE" },
    { accessToken }
  );
}
